File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/dialogs/mainDialog.js
/* eslint-disable jsdoc/require-returns */
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityTypes, InputHints, MessageFactory } = require('botbuilder');
const {
    ChoicePrompt,
    ChoiceFactory,
    ComponentDialog,
    DialogSet,
    WaterfallDialog,
    DialogTurnStatus,
    ListStyle,
} = require('botbuilder-dialogs');
const { RootBot } = require('../bots/rootBot');
const { SsoDialog } = require('./sso/ssoDialog');
const { TangentDialog } = require('./tangentDialog');

const MAIN_DIALOG = 'MainDialog';
const FEEDBACK_PROMPT = 'FeedbackPrompt';
const CONFIRMATION_PROMPT = 'ConfirmationPrompt';
const TANGENT_DIALOG = 'TangentDialog';
const WATERFALL_DIALOG = 'WaterfallDialog';
const SSO_DIALOG_PREFIX = 'Sso';

class MainDialog extends ComponentDialog {
    /**
     * @param {import('botbuilder').ConversationState} conversationState
     * @param {import('../skillsConfiguration').SkillsConfiguration} skillsConfig
     * @param {import('botbuilder').SkillHttpClient} skillClient
     * @param {import('../skillConversationIdFactory').SkillConversationIdFactory} conversationIdFactory
     */
    constructor(conversationState, skillsConfig, skillClient, conversationIdFactory) {
        super(MAIN_DIALOG);

        const botId = process.env.MicrosoftAppId;

        if (!conversationState) throw new Error("[MainDialog]: Missing parameter 'conversationState' is required");
        if (!skillsConfig) throw new Error("[MainDialog]: Missing parameter 'skillsConfig' is required");
        if (!skillClient) throw new Error("[MainDialog]: Missing parameter 'skillClient' is required");
        if (!conversationIdFactory)
            throw new Error("[MainDialog]: Missing parameter 'conversationIdFactory' is required");

        this.feedbackProperty = conversationState.createProperty(RootBot.feedbackPropertyName);
        this.activeSkillProperty = conversationState.createProperty(RootBot.ActiveSkillPropertyName);
        this.skillsConfig = skillsConfig;
        this.deliveryMode = '';

        // Register the tangent dialog for testing tangents and resume.
        this.addDialog(new TangentDialog(TANGENT_DIALOG));

        // Add ChoicePrompt to render available delivery modes.
        this.addDialog(new ChoicePrompt(FEEDBACK_PROMPT));

        // Add ChoicePrompt to render available groups of skills.
        this.addDialog(new ChoicePrompt(CONFIRMATION_PROMPT));

        this.addDialog(
            new WaterfallDialog(WATERFALL_DIALOG, [
                this.askFeedbackStep.bind(this),
                this.ConfirmationStep.bind(this),
                this.finalStep.bind(this),
            ])
        );

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async run(turnContext, accessor) {
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this);

        const dialogContext = await dialogSet.createContext(turnContext);
        const results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dialogContext.beginDialog(this.id);
        }
    }

    /**
     * @param {import('botbuilder-dialogs').DialogContext} innerDc
     */
    async onContinueDialog(innerDc) {
        const activeSkill = await this.activeSkillProperty.get(innerDc.context, () => null);
        const activity = innerDc.context.activity;
        if (
            activeSkill != null &&
            activity.type === ActivityTypes.Message &&
            activity.text != null &&
            activity.text.toLowerCase() === 'abort'
        ) {
            // Cancel all dialogs when the user says abort.
            // The SkillDialog automatically sends an EndOfConversation message to the skill to let the
            // skill know that it needs to end its current dialogs, too.
            await innerDc.cancelAllDialogs();
            return innerDc.replaceDialog(this.initialDialogId, {
                text: 'Canceled! \n\n What delivery mode would you like to use?',
            });
        }
        return super.onContinueDialog(innerDc);
    }

    async askFeedbackStep(stepContext) {
        const CHOICES = ['⭐', '⭐⭐', '⭐⭐⭐'];

        return await stepContext.prompt(FEEDBACK_PROMPT, {
            choices: ChoiceFactory.toChoices(CHOICES),
            style: ListStyle.suggestedAction,
        });
    }

    async ConfirmationStep(stepContext) {
        this.feedback = stepContext.result.value;

        const messageText = `You gave us a ${this.feedback} rate. Is that correct?`;
        const retryMessageText = 'That was not a valid choice, please select a valid option.';

        // Create the PromptOptions from the skill configuration which contains the list of configured skills.
        return stepContext.prompt(CONFIRMATION_PROMPT, {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(retryMessageText, retryMessageText, InputHints.ExpectingInput),
            choices: ChoiceFactory.toChoices(['Yes', 'No']),
        });
    }

    async finalStep(stepContext) {
        const confirmation = stepContext.result.value;
        let messageText = '';

        if (confirmation == 'Yes') {
            messageText = 'Thank you for your feedback!';
        } else {
            messageText = 'Thank you for your feedback!';
        }
        await stepContext.context.sendActivity(messageText);

        // Restart dialog
        return stepContext.replaceDialog(this.initialDialogId);
    }
}

module.exports.MainDialog = MainDialog;

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/MainDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs.Sso;
using Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Skills;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs
{
    /// <summary>
    /// The main dialog for this bot. It uses a <see cref="SkillDialog"/> to call skills.
    /// </summary>
    public class MainDialog : ComponentDialog
    {
        // State property key that stores the active skill (used in AdapterWithErrorHandler to terminate the skills on error).
        public static readonly string ActiveSkillPropertyName = $"{typeof(MainDialog).FullName}.ActiveSkillProperty";

        private const string SsoDialogPrefix = "Sso";
        private readonly IStatePropertyAccessor<BotFrameworkSkill> _activeSkillProperty;
        private readonly string _deliveryMode = $"{typeof(MainDialog).FullName}.DeliveryMode";
        private readonly string _selectedSkillKey = $"{typeof(MainDialog).FullName}.SelectedSkillKey";
        private readonly SkillsConfiguration _skillsConfig;
        private readonly IConfiguration _configuration;

        // Dependency injection uses this constructor to instantiate MainDialog.
        public MainDialog(ConversationState conversationState, SkillConversationIdFactoryBase conversationIdFactory, SkillHttpClient skillClient, SkillsConfiguration skillsConfig, IConfiguration configuration)
            : base(nameof(MainDialog))
        {
            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));

            var botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;

            _skillsConfig = skillsConfig ?? throw new ArgumentNullException(nameof(skillsConfig));

            if (skillClient == null)
            {
                throw new ArgumentNullException(nameof(skillClient));
            }

            if (conversationState == null)
            {
                throw new ArgumentNullException(nameof(conversationState));
            }

            // Create state property to track the active skill.
            _activeSkillProperty = conversationState.CreateProperty<BotFrameworkSkill>(ActiveSkillPropertyName);

            // Add ChoicePrompt to render available delivery modes.
            AddDialog(new ChoicePrompt("FeedbackPrompt"));

            AddDialog(new ChoicePrompt("CheckAnswer"));

            // Add main waterfall dialog for this bot.
            var waterfallSteps = new WaterfallStep[]
            {
                SelectDeliveryModeStepAsync,
                CheckAnswer
            };
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), waterfallSteps));
            
            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> SelectDeliveryModeStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var messageText = "give feedback";
            const string retryMessageText = "invalid";
            var choices = new List<Choice>
            {
                new Choice("😀"),
                new Choice("⭐")
            };
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(retryMessageText, retryMessageText, InputHints.ExpectingInput),
                Choices = choices
            };

            // Prompt the user to select a delivery mode.
            return await stepContext.PromptAsync("FeedbackPrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> CheckAnswer(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var response = ((FoundChoice)stepContext.Result).Value;
            var messageText = $"Is the {response} correct?";
            const string retryMessageText = "invalid";
            var choices = new List<Choice>
            {
                new Choice("yes"),
                new Choice("no")
            };
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(retryMessageText, retryMessageText, InputHints.ExpectingInput),
                Choices = choices
            };

            return await stepContext.PromptAsync("CheckAnswer", options, cancellationToken);
        }
    }
}
